flocker/node/agents/test/test_blockdevice.py

```diff
@@ -10,18 +10,401 @@
+            ``BlockDeviceDeployer``.
+        :param local_state: As accepted by
+            ``IDeployer.calculate_necessary_state_changes``.
+        :param desired_configuration: As accepted by
+            ``IDeployer.calculate_necessary_state_changes``.
+
+        :return: The return value of
+            ``BlockDeviceDeployer.calculate_necessary_state_changes``.
+        """
+        # Control service still reports that this node has no manifestations.
+        current_cluster_state = Deployment(
+            nodes=frozenset([Node(hostname=local_hostname)])
+        )
+
+        api = LoopbackBlockDeviceAPI.from_path(self.mktemp())
+        deployer = BlockDeviceDeployer(
+            hostname=local_hostname,
+            block_device_api=api,
+        )
+
+        return deployer.calculate_necessary_state_changes(
+            local_state, desired_configuration, current_cluster_state
+        )
+
+    def test_local_state_overrides_cluster_state(self):
+        """
+        ``BlockDeviceDeployer.calculate_necessary_state_changes`` bases its
+        decision about whether it is necessary to create a dataset by
+        inspecting the ``local_state`` argument.  It does this, instead of
+        inspecting the ``current_cluster_state`` argument, because the
+        ``local_state`` is presumed to be the most up-to-date information.
+        ``current_cluster_state`` has traveled a round-trip between the node
+        and the control service and had that period of time to fall behind the
+        actual node state.
+        """
+        local_hostname = u"192.0.2.1"
+        dataset_id = unicode(uuid4())
+        dataset = Dataset(dataset_id=dataset_id)
+        manifestation = Manifestation(
+            dataset=dataset, primary=True
+        )
+
+        # Discovered local state reveals the configured manifestation is
+        # already on this node.
+        local_state = NodeState(
+            hostname=local_hostname,
+            manifestations=[manifestation],
+            paths={dataset_id: FilePath('/foo/bar')}
+        )
+
+        # Configuration requires a manifestation on this node.
+        desired_configuration = Deployment(
+            nodes=frozenset({
+                Node(
+                    hostname=local_hostname,
+                    manifestations={dataset_id: manifestation},
+                )
+            })
+        )
+
+        actual_changes = self._calculate_changes(
+            local_hostname, local_state, desired_configuration
+        )
+
+        # If Deployer is buggy and not overriding cluster state with local
+        # state this would result in a dataset creation action:
+        expected_changes = InParallel(changes=[])
+
+        self.assertEqual(expected_changes, actual_changes)
+
+    def test_match_configuration_to_state_of_datasets(self):
+        """
+        ``BlockDeviceDeployer.calculate_necessary_state_changes`` determines
```

Comment:
I don't think this is a good generic test; the description should probably be changed to specifically talk about metadata. A different size, in particular, would result in an action, so this is purely about metadata.